### ISSUE TITLE
Request for Hyperparameter Details for Reproducibility

### ISSUE THREAD

**Issue by heebinYoo** [kairanzhao/RUM#2]
First of all, I sincerely appreciate your contribution to NeurIPS 2024.
I have a question regarding hyperparameters for reproducibility, similar to [Issue #1](https://github.com/kairanzhao/RUM/issues/1).

In the OpenReview version of your paper ([link](https://openreview.net/pdf?id=QAbhLBF72K)), you have kindly provided many details in Appendix A.2. 
However, in the "Training details for machine unlearning" paragraph, the learning rates and other regularization parameters are only given as ranges.

Exhaustively searching through all values within these ranges is computationally expensive for me.
Additionally, after experimenting with several values from these ranges, I obtained significantly lower performance than reported in the paper, suggesting that reproducing the results without a thorough line search may not be feasible.

Would it be possible for you to share a table of the exact hyperparameters used for the reported results on the CIFAR-10, CIFAR-100, and Tiny-ImageNet datasets with ResNet-18, ResNet-50, and VGG models?
Alternatively, could you provide some guidance on the search process (assuming it was not manually tuned value by value)?

This would be extremely helpful for my research, as I plan to build upon your work and cite your paper.
I greatly appreciate your time and consideration.

Best regards,

HEE BIN YOO @ Seoul National University, South Korea

**Comment by kairanzhao** [comment]
Thank you for your interest and for testing our work. 

Regarding the hyperparameter details: our paper involves a wide range of experiments across multiple settings, and for many of the baselines, we followed the original publications’ hyperparameter tuning guidelines. If you could specify which particular experiment(s) you’re having trouble with, I’d be happy to offer more targeted advice. 

In general, please note that performance can be sensitive to even slight changes in hyperparameters or experimental setup, so additional careful tuning on your end may be necessary.

**Comment by heebinYoo** [comment]
Thank you so much for your response.
I was hesitant about whether to continue researching this because the hyperparameters seemed sensitive as you mentioned, but I gathered the courage to ask—and I’m really glad to get your response!

At the very least, if I have the hyperparameters that can reproduce the baseline results, I can verify whether I have set up my environment correctly, executed the code properly, and so on, allowing me to continue my experiments.
Could you provide the hyperparameters for running the settings in Table 1 of your paper—Fine-tune vanilla, L1-sparse vanilla, NegGrad+ vanilla, SalUn vanilla, and RUM—on CIFAR-10 and CIFAR-100?
In other words, could you share the missing hyperparameters (??) in the following commands?
Thank you in advance.



`python main_forget.py --unlearn FT --num_indexes_to_replace 3000 --unlearn_epochs 10 --dataset cifar10 --arch resnet18 --data /data/hbyoo/data --epochs 30 --unlearn_lr ??`
`python main_forget.py --unlearn FT --num_indexes_to_replace 3000  --unlearn_epochs 10 --dataset cifar100 --arch resnet50 --data /data/hbyoo/data --epochs 150 --unlearn_lr ??`

`python main_forget.py --unlearn FT_prune --num_indexes_to_replace 3000 --unlearn_epochs 10 --dataset cifar10 --arch resnet18 --data /data/hbyoo/data --epochs 30 --alpha ?? --unlearn_lr ?? `
`python main_forget.py --unlearn FT_prune --num_indexes_to_replace 3000 --unlearn_epochs 10 --dataset cifar100 --arch resnet50 --data /data/hbyoo/data --epochs 150 --alpha ??  --unlearn_lr ??`

`python main_forget.py --unlearn NG --num_indexes_to_replace 3000  --unlearn_epochs 5 --dataset cifar10 --arch resnet18 --data /data/hbyoo/data --epochs 30 --alpha ?? --unlearn_lr ??`
`python main_forget.py --unlearn NG --num_indexes_to_replace 3000 --unlearn_epochs 5 --dataset cifar100 --arch resnet50 --data /data/hbyoo/data --epochs 150 --alpha ?? --unlearn_lr ??`

`python main_random.py --unlearn RL --num_indexes_to_replace 3000 --batch_size 256 --dataset cifar10 --arch resnet18 --epochs 30 --data /data/hbyoo/data --path /data/hbyoo/RUM/assets/saliency_map/hard_??  --unlearn_lr ?? --unlearn_epochs ??`
`python main_random.py --unlearn RL --num_indexes_to_replace 3000 --batch_size 256 --dataset cifar100 --arch resnet50 --epochs 150 --data /data/hbyoo/data --path /data/hbyoo/RUM/assets/saliency_map/hard_??   --unlearn_lr ?? --unlearn_epochs ??`

`python main_rum.py --mem mix --num_indexes_to_replace 3000 --dataset cifar10 --arch resnet18 --epochs 30  --batch_size 256 --lr ??`
`python main_rum.py --mem mix --num_indexes_to_replace 3000 --dataset cifar100 --arch resnet50 --epochs 150 --batch_size 256 --lr ?? `

**Comment by kairanzhao** [comment]
Please note that Table 1 involves dozens of experiments, and we have provided tuning ranges for key hyperparameters in the paper. Could you clarify which specific experiment you're having trouble reproducing? This will help me offer more targeted support.

Thanks

**Comment by heebinYoo** [comment]
Thank you so much for your reply. As shown in the red box in the attached image, I would like to reproduce the ToW and MIA-gap for the vanilla version of each methodology. That’s what I was asking about the hyperparameters.

As I mentioned earlier, I appreciated and acknowledged that you provided a range, but I was wondering if you could provide the exact hyperparameters used for those specific experiments (the exact numbers for the parts marked with ?? in my previous response).

If this is considered a broad, untargeted question, I’ll try conducting a more specific line search first and come back to ask about the exact issues if they don’t work out in my situations. 


Thanks!

<img width="575" alt="Image" src="https://github.com/user-attachments/assets/27967b32-9388-4c1a-b87b-c6f1e472b7f5" />

**Comment by heebinYoo** [comment]
Thank you for your cooperation so far! I just performed a linear (or logarithmic) search for lr and alpha using a pure Bash script, as shown below:

`
lr_values=( $(python -c "import numpy as np; print(' '.join(map(str, np.logspace(-5, -1, num=10))))") )
for lr in "${lr_values[@]}"; do
    CUDA_VISIBLE_DEVICES=0 python main_forget.py --unlearn FT --num_indexes_to_replace 3000 --unlearn_epochs 10 \
        --dataset cifar10 --arch resnet18 --data /data/hbyoo/data --epochs 30 --unlearn_lr "$lr"
done
`
Everything has been successfully reproduced, and I will document this for future users of this repository before closing this issue.

Below are the rough reproduced results and the corresponding hyperparameters:
<img width="1180" alt="Image" src="https://github.com/user-attachments/assets/1e01c38c-5c5d-4c97-9d47-2cef84c58d2a" />

Thanks again for your support! 

I hope to see you at a conference someday!

**Comment by kairanzhao** [comment]
Thank you for sharing your detailed update and results. Your reproduction looks good, and I'm glad everything is running successfully. Just have one quick question: could you confirm whether you composed your forget set exactly as described in our paper—that is, including examples with mixed memorization levels? I noticed a slight difference in the reported forget accuracy compared to our numbers, though it may be due to other factors as well.

**Comment by heebinYoo** [comment]
In order to make the code easier to read, I performed the experiment without setting anything after removing the mem part. I fixed the seed to 2 (of course, other seeds also gave similar values). I used the forget set and retain set as they were generated in the code. Does other methods than the RUM, such as Fine-tune and L1-sparse, also need settings like mixed memorization levels to reproduce?

Maybe that could be the cause of the performance difference, as you predicted.

Firstly, I finished the reproduction and stopped the experiment for a while because of the ICML review.

If you reply, I will try that method as well.

Thank you.
